src/RobustAdaptiveMetropolis.jl

```diff
@@ -0,0 +1,205 @@
+
+julia> # Define a Gaussian with zero mean and some covariance.
+       struct Gaussian{A}
+           Σ::A
+       end
+
+julia> # Implement the LogDensityProblems interface.
+       LogDensityProblems.dimension(model::Gaussian) = size(model.Σ, 1)
+
+julia> function LogDensityProblems.logdensity(model::Gaussian, x)
+           d = LogDensityProblems.dimension(model)
+           return logpdf(MvNormal(zeros(d),model.Σ), x)
+       end
+
+julia> LogDensityProblems.capabilities(::Gaussian) = LogDensityProblems.LogDensityOrder{0}()
+
+julia> # Construct the model. We'll use a correlation of 0.5.
+       model = Gaussian([1.0 0.5; 0.5 1.0]);
+
+julia> # Number of samples we want in the resulting chain.
+       num_samples = 10_000;
+
+julia> # Number of warmup steps, i.e. the number of steps to adapt the covariance of the proposal.
+       # Note that these are not included in the resulting chain, as `discard_initial=num_warmup`
+       # by default in the `sample` call. To include them, pass `discard_initial=0` to `sample`.
+       num_warmup = 10_000;
+
+julia> # Set the seed so get some consistency.
+       Random.seed!(1234);
+
+julia> # Sample!
+       chain = sample(model, RAM(), 10_000; chain_type=Chains, num_warmup=10_000, progress=false, initial_params=zeros(2));
+
+julia> norm(cov(Array(chain)) - [1.0 0.5; 0.5 1.0]) < 0.2
+true
+```
+
+# References
+[^VIH12]: Vihola (2012) Robust adaptive Metropolis algorithm with coerced acceptance rate, Statistics and computing.
+"""
+Base.@kwdef struct RAM{T,A<:Union{Nothing,AbstractMatrix{T}}} <: AdvancedMH.MHSampler
+    "target acceptance rate"
+    α::T=0.234
+    "negative exponent of the adaptation decay rate"
+    γ::T=0.6
+    "initial lower-triangular Cholesky factor"
+    S::A=nothing
+    "lower bound on eigenvalues of the adapted covariance matrix"
+    eigenvalue_lower_bound::T=0.0
+    "upper bound on eigenvalues of the adapted covariance matrix"
+    eigenvalue_upper_bound::T=Inf
+end
+
+# TODO: Should we record anything like the acceptance rates?
+struct RAMState{T1,L,A,T2,T3}
+    x::T1
+    logprob::L
+    S::A
+    logα::T2
+    η::T3
+    iteration::Int
+    isaccept::Bool
+end
+
+AbstractMCMC.getparams(state::RAMState) = state.x
+AbstractMCMC.setparams!!(state::RAMState, x) = RAMState(x, state.logprob, state.S, state.logα, state.η, state.iteration, state.isaccept)
+
+function step_inner(
+    rng::Random.AbstractRNG,
+    model::AbstractMCMC.LogDensityModel,
+    sampler::RAM,
+    state::RAMState
+)
+    # This is the initial state.
+    f = model.logdensity
+    d = LogDensityProblems.dimension(f)
+
+    # Sample the proposal.
+    x = state.x
+    U = randn(rng, d)
+    x_new = x + state.S * U
+
+    # Compute the acceptance probability.
+    lp = state.logprob
+    lp_new = LogDensityProblems.logdensity(f, x_new)
+    logα = min(lp_new - lp, zero(lp))  # `min` because we'll use this for updating
```

Comment:
But IMO this makes it a bit strange if we then put the _unbounded_ `logα` in the resulting state, since this is _not_ the quantity used to update the `S` :confused: 

And for the purposes it is user here, it doesn't actually matter if it's bounded or not, right? As in, it's equivalent here, but not equivalent in the `S` update, hence it seems somewhat natural for me to just do it once and for all.

Reply:
> since this is not the quantity used to update the S

It is, isn't it? The update is just slightly different. Otherwise with the same reasoning you could also argue that one should only store `α` (or maybe the difference to the targeted α?) since only these are used to update S.

In the end I guess it doesn't matter as it's only used in these two places. It just felt strange conceptually to bound it here, in particular since it seemed you already felt the need to explain this decision with a comment.

Reply:
But alpha represents the acceptance probability, no? So clamping it like this is _technically_ always what you should do, but most of time we don't because it's unnecessary for sampling according to this probability.

_However_, if the user wants to actually look at the resulting acceptance probs, then it's a question of: do we want the user to do
```julia
mean(exp, getproperty.(states, :logα)))
```
or
```julia
mean(exp, min.(1, getproperty.(states, :logα)))
```
In my head, the user _expects_ to do the former, not the latter 🤷

Reply:
I can see your point. Even though I think neither of the two alternatives is particularly user-friendly, IMO a separate API for acceptance probabilities would be better.

In any case, I think I wouldn't even have commented on this line if the comment `# `min` because we'll use this for updating` would not have been there. So maybe just remove the comment?

Reply:
Fair, but then I'm worried someone might come along later and go "wait, that's not needed; let's just remove this unnecessary `min`", not realizing that we'll use this for adaptation 😬

Reply:
Maybe add an `@assert logα <= zero(logα)` at the top of `ram_adapt`?